src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -332,15 +337,27 @@ static void verifyColumnMasterKeyMetadata(SQLServerConnection connection, SQLSer
     static SqlAuthenticationToken getManagedIdentityCredAuthToken(String resource,
             String managedIdentityClientId) throws SQLServerException {
         ManagedIdentityCredential mic = null;
+        String key = managedIdentityClientId + ManagedIdentityCredential.class.getSimpleName();
 
         if (logger.isLoggable(java.util.logging.Level.FINEST)) {
             logger.finest("Getting Managed Identity authentication token for: " + managedIdentityClientId);
         }
 
         if (null != managedIdentityClientId && !managedIdentityClientId.isEmpty()) {
-            mic = new ManagedIdentityCredentialBuilder().clientId(managedIdentityClientId).build();
+            mic = (ManagedIdentityCredential) getCredentialFromCache(key);
+
+            if (null == mic) {

```

Comment:
If parallel operations are happening, this could still result in duplicate credential objects in use (even though only the last one is cached). Adding a new credential to the cache should be an atomic operation with a lock to ensure we only ever create one for each key.